man/inotifywait.1
.TH inotifywait 1 "_DATE_" "inotifywait _PACKAGE_VERSION_"

.SH NAME
inotifywait \- wait for changes to files using inotify

.SH SYNOPSIS
.B inotifywait
.RB [ \-hcmrq ]
.RB [ \-e
<event> ]
.RB [ \-t
<seconds> ]
.RB [ \-\-format
<fmt> ]
.RB [ \-\-timefmt
<fmt> ]
<file> [ ... ]

.SH DESCRIPTION
.B inotifywait
efficiently waits for changes to files using Linux's
.BR inotify(7)
interface.  It is suitable for waiting for changes to files from shell scripts.
It can either exit once an event occurs, or continually execute and output events
as they occur.

.SH OUTPUT
.B inotifywait
will output diagnostic information on standard error and event information on
standard output.  The event output can be configured, but by default it
consists of lines of the following form:

.B watched_filename EVENT_NAMES event_filename

.TP
.B watched_filename
is the name of the file on which the event occurred.  If the file is a directory,
a trailing slash is output.
.TP
.B EVENT_NAMES
are the names of the inotify events which occurred, separated by commas.
.TP
.B event_filename
is output only when the event occurred on a directory, and in this case the name
of the file within the directory which caused this event is output.

By default, any special characters in filenames are not escaped in any way.  This
can make the output of inotifywait difficult to parse in awk scripts or similar.
The
.B \-\-csv
and
.B \-\-format
options will be helpful in this case.

.SH OPTIONS
.TP
.B \-h, \-\-help
Output some helpful usage information.
.TP
.B @<file>
When watching a directory tree recursively, exclude the specified file from
being watched.  The file must be specified with a relative or absolute path
according to whether a relative or absolute path is given for watched
directories.  If a specific path is explicitly both included and excluded, it
will always be watched.

.B Note:
If you need to watch a directory or file whose name starts with @, give the
absolute path.
.TP
.B \-\-fromfile <file>
Read filenames to watch or exclude from a file, one filename per line.  If
filenames begin with @ they are excluded as described above.  If <file> is `-',
filenames are read from standard input.  Use this option if you need to watch
too many files to pass in as command line arguments.
.TP
.B \-m, \-\-monitor
Instead of exiting after receiving a single event, execute indefinitely.  The
default behaviour is to exit after the first event occurs.
.TP
.B \-d, \-\-daemon
Same as \-\-monitor, except run in the background logging events to a file
that must be specified by \-\-outfile. Implies \-\-syslog.
.TP
.B \-o, \-\-outfile <file>
Output events to <file> rather than stdout.
.TP
.B \-s, \-\-syslog
Output errors to
.BR syslog(3)
system log module rather than stderr.
.TP
.B \-r, \-\-recursive
Watch all subdirectories of any directories passed as arguments.  Watches
will be set up recursively to an unlimited depth.  Symbolic links are not
traversed.  Newly created subdirectories will also be watched.

.B Warning:
If you use this option while watching the root directory
of a large tree, it may take quite a while until all inotify watches are
established, and events will not be received in this time.  Also, since one
inotify watch will be established per subdirectory, it is possible that the
maximum amount of inotify watches per user will be reached.  The default
maximum is 8192; it can be increased by writing to
.BR /proc/sys/fs/inotify/max_user_watches .

.TP
.B \-q, \-\-quiet
If specified once, the program will be less verbose.  Specifically, it will not
state when it has completed establishing all inotify watches.

If specified twice, the program will output nothing at all, except in the case
of fatal errors.

.TP
.B \-\-exclude <pattern>
Do not process any events whose filename matches the specified POSIX extended
regular expression, case sensitive.

.TP
.B \-\-excludei <pattern>
Do not process any events whose filename matches the specified POSIX extended
regular expression, case insensitive.

.TP
.B \-t <seconds>, \-\-timeout <seconds>
Exit if an appropriate event has not occurred within <seconds> seconds. If
<seconds> is zero (the default), wait indefinitely for an event.

.TP
.B \-e <event>, \-\-event <event>
Listen for specific event(s) only.  The events which can be listened for are
listed in the
.B EVENTS
section.  This option can be specified more than once.  If omitted, all events
are listened for.

.TP
.B \-c, \-\-csv
Output in CSV (comma-separated values) format.  This is useful when filenames
may contain spaces, since in this case it is not safe to simply split the output
at each space character.

.TP
.B \-\-timefmt <fmt>
Set a time format string as accepted by strftime(3) for use with the `%T' conversion
in the \-\-format option.

.TP
.B \-\-format <fmt>
Output in a user-specified format, using printf-like syntax.  The event strings
output are limited to around 4000 characters and will be truncated to this length.
The following conversions are supported:

.TP
%w
This will be replaced with the name of the Watched file on which an event occurred.

.TP
%f
When an event occurs within a directory, this will be replaced with the name of the
File which caused the event to occur.  Otherwise, this will be replaced with an
empty string.

.TP
%e
Replaced with the Event(s) which occurred, comma-separated.

.TP
%Xe
Replaced with the Event(s) which occurred, separated by whichever character is
in the place of `X'.

.TP
%T
Replaced with the current Time in the format specified by the \-\-timefmt option,
which should be a format string suitable for passing to strftime(3).

.TP
.B \-\-execute <program>
Execute the specified program when an event occurs.  argv[1] will be the full
path to the event.



.SH "EXIT STATUS"
.TP
.B 0
The program executed successfully, and an event occurred which was being
listened for.
.TP
.B 1
An error occurred in execution of the program, or an event occurred which was
not being listened for.  The latter generally occurs if something happens which
forcibly removes the inotify watch, such as a watched file being deleted or the
filesystem containing a watched file being unmounted.
.TP
.B 2
The
.B \-t
option was used and an event did not occur in the specified interval of time.

.SH EVENTS
The following events are valid for use with the
.B \-e
option:

.TP
.B access
A watched file or a file within a watched directory was read from.

.TP
.B modify
A watched file or a file within a watched directory was written to.

.TP
.B attrib
The metadata of a watched file or a file within a watched directory was
modified.  This includes timestamps, file permissions, extended attributes etc.

.TP
.B close_write
A watched file or a file within a watched directory was closed, after being
opened in writeable mode.  This does not necessarily imply the file was written
to.

.TP
.B close_nowrite
A watched file or a file within a watched directory was closed, after being
opened in read-only mode.

.TP
.B close
A watched file or a file within a watched directory was closed, regardless of
how it was opened.  Note that this is actually implemented simply by listening
for both
.B close_write
and
.B close_nowrite,
hence all close events received will be output as one of these, not
.B CLOSE.

.TP
.B open
A watched file or a file within a watched directory was opened.

.TP
.B moved_to
A file or directory was moved into a watched directory.  This event occurs even
if the file is simply moved from and to the same directory.

.TP
.B moved_from
A file or directory was moved from a watched directory.  This event occurs even
if the file is simply moved from and to the same directory.

.TP
.B move
A file or directory was moved from or to a watched directory.  Note that this is
actually implemented simply by listening for both
.B moved_to
and
.B moved_from,
hence all close events received will be output as one or both of these, not
.B MOVE.

.TP
.B move_self
A watched file or directory was moved. After this event, the file or directory 
is no longer being watched.

.TP
.B create
A file or directory was created within a watched directory.

.TP
.B delete
A file or directory within a watched directory was deleted.

.TP
.B delete_self
A watched file or directory was deleted.  After this event the file or directory
is no longer being watched.  Note that this event can occur even if it is not
explicitly being listened for.

.TP
.B unmount
The filesystem on which a watched file or directory resides was unmounted.
After this event the file or directory is no longer being watched.  Note that
this event can occur even if it is not explicitly being listened to.


.SH EXAMPLES

.SS Example 1
Running inotifywait at the command-line to wait for any file in the `test'
directory to be accessed.  After running inotifywait, `cat test/foo' is run
in a separate console.

.nf
% inotifywait test
Setting up watches.
Watches established.
test/ ACCESS foo
.fi

.SS Example 2
A short shell script to efficiently wait for httpd-related log messages and
do something appropriate.

.nf
#!/bin/sh
while inotifywait -e modify /var/log/messages; do
  if tail -n1 /var/log/messages | grep httpd; then
    kdialog --msgbox "Apache needs love!"
  fi
done
.fi

.SS Example 3
A custom output format is used to watch `~/test'.  Meanwhile, someone runs
`touch ~/test/badfile; touch ~/test/goodfile; rm ~/test/badfile' in another
console.

.nf
% inotifywait -m -r --format '%:e %f' ~/test
Setting up watches.  Beware: since -r was given, this may take a while!
Watches established.
CREATE badfile
OPEN badfile
ATTRIB badfile
CLOSE_WRITE:CLOSE badfile
CREATE goodfile
OPEN goodfile
ATTRIB goodfile
CLOSE_WRITE:CLOSE goodfile
DELETE badfile
.fi


.SH BUGS
There are race conditions in the recursive directory watching code
which can cause events to be missed if they occur in a directory immediately
after that directory is created.  This is probably not fixable.

It is assumed the inotify event queue will never overflow.

.SH AUTHORS
inotifywait is written and maintained by Rohan McGovern <rohan@mcgovern.id.au>.

inotifywait is part of inotify-tools.  The inotify-tools website is located at:
.I http://inotify-tools.sourceforge.net/

.SH "SEE ALSO"
inotifywatch(1), strftime(3), inotify(7)


NEWS

 * 19 October 2005
   inotify-tools begun.
   


src/inotifywait.c
#include "../config.h"
#include "common.h"

#include <sys/types.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <regex.h>
#include <signal.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <syslog.h>
#include <unistd.h>

#include <inotifytools/inotifytools.h>
#include <inotifytools/inotify.h>

extern char *optarg;
extern int optind, opterr, optopt;

#define MAX_STRLEN 4096
#define EXCLUDE_CHUNK 1024

#define nasprintf(...) niceassert( -1 != asprintf(__VA_ARGS__), "out of memory")

// METHODS
bool parse_opts(
  int * argc,
  char *** argv,
  int * events,
  bool * monitor,
  int * quiet,
  unsigned long int * timeout,
  int * recursive,
  bool * csv,
  bool * daemon,
  bool * syslog,
  bool * verbose,
  char ** format,
  char ** timefmt,
  char ** fromfile,
  char ** outfile,
  char ** regex,
  char ** iregex,
  char ** execute
);

void print_help();


char * csv_escape( char * string ) {
	static char csv[MAX_STRLEN+1];
	static unsigned int i, ind;

	if ( strlen(string) > MAX_STRLEN ) {
		return NULL;
	}

	if ( strlen(string) == 0 ) {
		return NULL;
	}

	// May not need escaping
	if ( !strchr(string, '"') && !strchr(string, ',') && !strchr(string, '\n')
	     && string[0] != ' ' && string[strlen(string)-1] != ' ' ) {
		strcpy( csv, string );
		return csv;
	}

	// OK, so now we _do_ need escaping.
	csv[0] = '"';
	ind = 1;
	for ( i = 0; i < strlen(string); ++i ) {
		if ( string[i] == '"' ) {
			csv[ind++] = '"';
		}
		csv[ind++] = string[i];
	}
	csv[ind++] = '"';
	csv[ind] = '\0';

	return csv;
}


void validate_format( char * fmt ) {
	// Make a fake event
	struct inotify_event * event =
	   (struct inotify_event *)malloc(sizeof(struct inotify_event) + 4);
	if ( !event ) {
		fprintf( stderr, "Seem to be out of memory... yikes!\n" );
		exit(EXIT_FAILURE);
	}
	event->wd = 0;
	event->mask = IN_ALL_EVENTS;
	event->len = 3;
	strcpy( event->name, "foo" );
	FILE * devnull = fopen( "/dev/null", "a" );
	if ( !devnull ) {
		fprintf( stderr, "Couldn't open /dev/null: %s\n", strerror(errno) );
		free( event );
		return;
	}
	if ( -1 == inotifytools_fprintf( devnull, event, fmt ) ) {
		fprintf( stderr, "Something is wrong with your format string.\n" );
		exit(EXIT_FAILURE);
	}
	free( event );
	fclose(devnull);
}


void output_event_csv( struct inotify_event * event ) {
    char *filename = csv_escape(inotifytools_filename_from_wd(event->wd));
    if (filename != NULL)
        printf("%s,", filename);

	printf("%s,", csv_escape( inotifytools_event_to_str( event->mask ) ) );
	if ( event->len > 0 )
		printf("%s", csv_escape( event->name ) );
	printf("\n");
}


void output_error( bool syslog, char* fmt, ... ) {
	va_list va;
	va_start(va, fmt);
	if ( syslog ) {
		vsyslog(LOG_INFO, fmt, va);
	} else {
		vfprintf(stderr, fmt, va);
	}
	va_end(va);
}


int main(int argc, char ** argv)
{
	int events = 0;
        int orig_events;
	bool monitor = false;
	int quiet = 0;
	unsigned long int timeout = 0;
	int recursive = 0;
	bool csv = false;
	bool dodaemon = false;
	bool syslog = false;
    bool verbose = false;
	char * format = NULL;
	char * timefmt = NULL;
	char * fromfile = NULL;
	char * outfile = NULL;
	char * regex = NULL;
	char * iregex = NULL;
	char * execute = NULL;
    int fd;

	// Parse commandline options, aborting if something goes wrong
	if ( !parse_opts(&argc, &argv, &events, &monitor, &quiet, &timeout,
	                 &recursive, &csv, &dodaemon, &syslog, &verbose, &format, &timefmt,
	                 &fromfile, &outfile, &regex, &iregex, &execute) ) {
		return EXIT_FAILURE;
	}

    if (verbose)
        fprintf(stderr, "Running verbosely. \n");

	if ( !inotifytools_initialize() ) {
		fprintf(stderr, "Couldn't initialize inotify.  Are you running Linux "
		                "2.6.13 or later, and was the\n"
		                "CONFIG_INOTIFY option enabled when your kernel was "
		                "compiled?  If so, \n"
		                "something mysterious has gone wrong.  Please e-mail "
		                PACKAGE_BUGREPORT "\n"
		                " and mention that you saw this message.\n");
		return EXIT_FAILURE;
	}

	if ( timefmt ) inotifytools_set_printf_timefmt( timefmt );
	if (
		(regex && !inotifytools_ignore_events_by_regex(regex, REG_EXTENDED) ) ||
		(iregex && !inotifytools_ignore_events_by_regex(iregex, REG_EXTENDED|
		                                                        REG_ICASE))
	) {
		fprintf(stderr, "Error in `exclude' regular expression.\n");
		return EXIT_FAILURE;
	}


	if ( format ) validate_format(format);

	// Attempt to watch file
	// If events is still 0, make it all events.
	if (events == 0)
		events = IN_ALL_EVENTS;
        orig_events = events;
        if ( monitor && recursive ) {
                events = events | IN_CREATE | IN_MOVED_TO | IN_MOVED_FROM;
        }

	FileList list = construct_path_list( argc, argv, fromfile );

	if (0 == list.watch_files[0]) {
		fprintf(stderr, "No files specified to watch!\n");
		return EXIT_FAILURE;
	}


    // Daemonize - BSD double-fork approach
	if ( dodaemon ) {
#ifdef HAVE_DAEMON
		if (daemon(0, 0)) {
			fprintf(stderr, "Failed to daemonize!\n");
			return EXIT_FAILURE;
		}
#else
		pid_t pid = fork();
	        if (pid < 0) {
			fprintf(stderr, "Failed to fork1 whilst daemonizing!\n");
	                return EXIT_FAILURE;
	        } 
	        if (pid > 0) {
			_exit(0);
	        }
		if (setsid() < 0) {
			fprintf(stderr, "Failed to setsid whilst daemonizing!\n");
	                return EXIT_FAILURE;
	        }
		signal(SIGHUP,SIG_IGN);
	        pid = fork();
	        if (pid < 0) {
	                fprintf(stderr, "Failed to fork2 whilst daemonizing!\n");
	                return EXIT_FAILURE;
	        }
	        if (pid > 0) {
	                _exit(0);
	        }
		if (chdir("/") < 0) {
			fprintf(stderr, "Failed to chdir whilst daemonizing!\n");
	                return EXIT_FAILURE;
	        }

		// Redirect stdin from /dev/null
	        fd = open("/dev/null", O_RDONLY);
		if (fd != fileno(stdin)) {
			dup2(fd, fileno(stdin));
			close(fd);
		}
#endif

		// Redirect stdout to a file
	        fd = open(outfile, O_WRONLY | O_CREAT | O_APPEND, 0600);
		if (fd < 0) {
                        fprintf( stderr, "Failed to open output file %s\n", outfile );
                        return EXIT_FAILURE;
                }
		if (fd != fileno(stdout)) {
			dup2(fd, fileno(stdout));
			close(fd);
		}

        // Redirect stderr to /dev/null
		fd = open("/dev/null", O_WRONLY);
	        if (fd != fileno(stderr)) {
	                dup2(fd, fileno(stderr));
	                close(fd);
	        }
	
        } else if (outfile != NULL) { // Redirect stdout to a file if specified
		fd = open(outfile, O_WRONLY | O_CREAT | O_APPEND, 0600);
		if (fd < 0) {
			fprintf( stderr, "Failed to open output file %s\n", outfile );
			return EXIT_FAILURE;
		}
		if (fd != fileno(stdout)) {
			dup2(fd, fileno(stdout));
			close(fd);
		}
        }

        if ( syslog ) {
		openlog ("inotifywait", LOG_CONS | LOG_PID | LOG_NDELAY, LOG_DAEMON);
        }

	if ( !quiet ) {
		if ( recursive ) {
			output_error( syslog, "Setting up watches.  Beware: since -r "
				"was given, this may take a while!\n" );
		} else {
			output_error( syslog, "Setting up watches.\n" );
		}
	}

	// now watch files
	for ( int i = 0; list.watch_files[i]; ++i ) {
		char const *this_file = list.watch_files[i];
		if ( (recursive && !inotifytools_watch_recursively_with_exclude(
		                        this_file,
		                        events,
		                        list.exclude_files ))
		     || (!recursive && !inotifytools_watch_file( this_file, events )) ){
			if ( inotifytools_error() == ENOSPC ) {
				output_error( syslog, "Failed to watch %s; upper limit on inotify "
				                "watches reached!\n", this_file );
				output_error( syslog, "Please increase the amount of inotify watches "
				        "allowed per user via `/proc/sys/fs/inotify/"
				        "max_user_watches'.\n");
                return EXIT_FAILURE;
			}
            else if ( inotifytools_error() == ENOENT ) {
                output_error( syslog, "Failed to watch '%s'; %s. continuing.\n",
                        this_file, strerror( inotifytools_error() ) );
            }
			else {
				output_error( syslog, "Couldn't watch %s: %s\n", this_file,
				        strerror( inotifytools_error() ) );
                return EXIT_FAILURE;
			}
		}
	}

	if ( !quiet ) {
		output_error( syslog, "Watches established.\n" );
	}

	// Now wait till we get event
	struct inotify_event * event;
	char * moved_from = 0;

	do {
		event = inotifytools_next_event( timeout );
		if ( !event ) {
			if ( !inotifytools_error() ) {
				return EXIT_TIMEOUT;
			}
			else {
				output_error( syslog, "%s\n", strerror( inotifytools_error() ) );
				return EXIT_FAILURE;
			}
		}

		if ( quiet < 2 && (event->mask & orig_events) ) {
			if ( csv ) {
				output_event_csv( event );
			}
			else if ( format ) {
				inotifytools_printf( event, format );
			}
			else {
				inotifytools_printf( event, "%w %,e %f\n" );
			}
		}

		// if we last had MOVED_FROM and don't currently have MOVED_TO,
		// moved_from file must have been moved outside of tree - so unwatch it.
		if ( moved_from && !(event->mask & IN_MOVED_TO) ) {
			if ( !inotifytools_remove_watch_by_filename( moved_from ) ) {
				output_error( syslog, "Error removing watch on %s: %s\n",
				         moved_from, strerror(inotifytools_error()) );
			}
			free( moved_from );
			moved_from = 0;
		}

		if ( monitor && recursive ) {
			if ((event->mask & IN_CREATE) ||
			    (!moved_from && (event->mask & IN_MOVED_TO))) {
				// New file - if it is a directory, watch it
				static char * new_file;

				nasprintf( &new_file, "%s%s",
				           inotifytools_filename_from_wd( event->wd ),
				           event->name );

				if ( isdir(new_file) &&
				    !inotifytools_watch_recursively( new_file, events ) ) {
					output_error( syslog, "Couldn't watch new directory %s: %s\n",
					         new_file, strerror( inotifytools_error() ) );
				}
				free( new_file );
			} // IN_CREATE
			else if (event->mask & IN_MOVED_FROM) {
				nasprintf( &moved_from, "%s%s/",
				           inotifytools_filename_from_wd( event->wd ),
				           event->name );
				// if not watched...
				if ( inotifytools_wd_from_filename(moved_from) == -1 ) {
					free( moved_from );
					moved_from = 0;
				}
			} // IN_MOVED_FROM
			else if (event->mask & IN_MOVED_TO) {
				if ( moved_from ) {
					static char * new_name;
					nasprintf( &new_name, "%s%s/",
					           inotifytools_filename_from_wd( event->wd ),
					           event->name );
					inotifytools_replace_filename( moved_from, new_name );
					free( moved_from );
					moved_from = 0;
				} // moved_from
			}
		}

		// If user wants us to run something, do it now
		if ( execute ) {
			pid_t pid = fork();
			if (pid == -1) {
				output_error( syslog, "Error forking child: %s\n",
				         strerror(errno) );
			} else if (pid == 0) {
				char *argv[] = { execute, inotifytools_filename_from_wd(event->wd), NULL, };
				exit(execvp(execute, argv));
			}
		}

		fflush( NULL );

	} while ( monitor );

	// If we weren't trying to listen for this event...
	if ( (events & event->mask) == 0 ) {
		// ...then most likely something bad happened, like IGNORE etc.
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


bool parse_opts(
  int * argc,
  char *** argv,
  int * events,
  bool * monitor,
  int * quiet,
  unsigned long int * timeout,
  int * recursive,
  bool * csv,
  bool * daemon,
  bool * syslog,
  bool * verbose,
  char ** format,
  char ** timefmt,
  char ** fromfile,
  char ** outfile,
  char ** regex,
  char ** iregex,
  char ** execute
) {
	assert( argc ); assert( argv ); assert( events ); assert( monitor );
	assert( quiet ); assert( timeout ); assert( csv ); assert( daemon );
	assert( syslog ); assert(verbose); assert( format ); assert( timefmt ); assert( fromfile ); 
	assert( outfile ); assert( regex ); assert( iregex );

	// Settings for options
	int new_event;
	char * timeout_end = NULL;
	// format with trailing newline
	static char * newlineformat;

	// Short options
	static const char opt_string[] = "mrhcdsvqt:fo:e:";

	// Long options
	static const struct option long_opts[] = {
		{"help",              no_argument, NULL, 'h'},
		{"event",       required_argument, NULL, 'e'},
		{"monitor",           no_argument, NULL, 'm'},
		{"quiet",             no_argument, NULL, 'q'},
		{"timeout",     required_argument, NULL, 't'},
		{"filename",          no_argument, NULL, 'f'},
		{"recursive",         no_argument, NULL, 'r'},
		{"csv",               no_argument, NULL, 'c'},
		{"daemon",            no_argument, NULL, 'd'},
		{"syslog",            no_argument, NULL, 's'},
        {"verbose",           no_argument, NULL, 'v'},
		{"format",      required_argument, NULL, 'n'},
		{"timefmt",     required_argument, NULL, 'i'},
		{"fromfile",    required_argument, NULL, 'z'},
		{"outfile",     required_argument, NULL, 'o'},
		{"exclude",     required_argument, NULL, 'a'},
		{"excludei",    required_argument, NULL, 'b'},
		{"execute",     required_argument, NULL, 'x'},
		{NULL, 0, 0, 0},
	};

	// Get first option
	char curr_opt = getopt_long(*argc, *argv, opt_string, long_opts, NULL);

	// While more options exist...
	while ( (curr_opt != '?') && (curr_opt != (char)-1) )
	{
		switch ( curr_opt )
		{
			// --help or -h
			case 'h':
				print_help();
				// Shouldn't process any further...
				return false;
				break;

			// --monitor or -m
			case 'm':
				*monitor = true;
				break;

			// --quiet or -q
			case 'q':
				(*quiet)++;
				break;

			// --recursive or -r
			case 'r':
				(*recursive)++;
				break;

			// --csv or -c
			case 'c':
				(*csv) = true;
				break;

			// --daemon or -d
			case 'd':
				(*daemon) = true;
				(*monitor) = true;
				(*syslog) = true;
				break;

			// --syslog or -s
			case 's':
				(*syslog) = true;
				break;

            // --verbose or -v
            case 'v':
                (*verbose) = true;
                break;

			// --filename or -f
			case 'f':
				fprintf(stderr, "The '--filename' option no longer exists.  "
				                "The option it enabled in earlier\nversions of "
				                "inotifywait is now turned on by default.\n");
				return false;
				break;

			// --format
			case 'n':
				newlineformat = (char *)malloc(strlen(optarg)+2);
				strcpy( newlineformat, optarg );
				strcat( newlineformat, "\n" );
				(*format) = newlineformat;
				break;

			// --timefmt
			case 'i':
				(*timefmt) = optarg;
				break;

			// --exclude
			case 'a':
				(*regex) = optarg;
				break;

			// --excludei
			case 'b':
				(*iregex) = optarg;
				break;

			// --fromfile
			case 'z':
				if (*fromfile) {
					fprintf(stderr, "Multiple --fromfile options given.\n");
					return false;
				}
				(*fromfile) = optarg;
				break;

			// --outfile
			case 'o':
				if (*outfile) {
					fprintf(stderr, "Multiple --outfile options given.\n");
					return false;
				}
				(*outfile) = optarg;
				break;

			// --timeout or -t
			case 't':
				*timeout = strtoul(optarg, &timeout_end, 10);
				if ( *timeout_end != '\0' )
				{
					fprintf(stderr, "'%s' is not a valid timeout value.\n"
					        "Please specify an integer of value 0 or "
					        "greater.\n",
					        optarg);
					return false;
				}
				break;

			// --event or -e
			case 'e':
				// Get event mask from event string
				new_event = inotifytools_str_to_event(optarg);

				// If optarg was invalid, abort
				if ( new_event == -1 )
				{
					fprintf(stderr, "'%s' is not a valid event!  Run with the "
					                "'--help' option to see a list of "
					                "events.\n", optarg);
					return false;
				}

				// Add the new event to the event mask
				(*events) = ( (*events) | new_event );

				break;

			// --execute
			case 'x':
				(*execute) = optarg;
				break;
		}

		curr_opt = getopt_long(*argc, *argv, opt_string, long_opts, NULL);

	}

	if ( *monitor && *timeout != 0 ) {
		fprintf(stderr, "-m and -t cannot both be specified.\n");
		return false;
	}

	if ( *regex && *iregex ) {
		fprintf(stderr, "--exclude and --excludei cannot both be specified.\n");
		return false;
	}

	if ( *format && *csv ) {
		fprintf(stderr, "-c and --format cannot both be specified.\n");
		return false;
	}

	if ( !*format && *timefmt ) {
		fprintf(stderr, "--timefmt cannot be specified without --format.\n");
		return false;
	}

	if ( *format && strstr( *format, "%T" ) && !*timefmt ) {
		fprintf(stderr, "%%T is in --format string, but --timefmt was not "
		                "specified.\n");
		return false;
	}

	if ( *daemon && *outfile == NULL ) {
		fprintf(stderr, "-o must be specified with -d.\n");
		return false;
	}

	(*argc) -= optind;
	*argv = &(*argv)[optind];

	// If ? returned, invalid option
	return (curr_opt != '?');
}


void print_help()
{
	printf("inotifywait %s\n", PACKAGE_VERSION);
	printf("Wait for a particular event on a file or set of files.\n");
	printf("Usage: inotifywait [ options ] file1 [ file2 ] [ file3 ] "
	       "[ ... ]\n");
	printf("Options:\n");
	printf("\t-h|--help     \tShow this help text.\n");
	printf("\t@<file>       \tExclude the specified file from being "
	       "watched.\n");
	printf("\t--exclude <pattern>\n"
	       "\t              \tExclude all events on files matching the\n"
	       "\t              \textended regular expression <pattern>.\n");
	printf("\t--excludei <pattern>\n"
	       "\t              \tLike --exclude but case insensitive.\n");
	printf("\t-m|--monitor  \tKeep listening for events forever.  Without\n"
	       "\t              \tthis option, inotifywait will exit after one\n"
	       "\t              \tevent is received.\n");
	printf("\t-d|--daemon   \tSame as --monitor, except run in the background\n"
               "\t              \tlogging events to a file specified by --outfile.\n"
               "\t              \tImplies --syslog.\n");
	printf("\t-r|--recursive\tWatch directories recursively.\n");
	printf("\t--fromfile <file>\n"
	       "\t              \tRead files to watch from <file> or `-' for "
	       "stdin.\n");
	printf("\t-o|--outfile <file>\n"
	       "\t              \tPrint events to <file> rather than stdout.\n");
	printf("\t-s|--syslog   \tSend errors to syslog rather than stderr.\n");
    printf("\t-v|--verbose  \tPrint more (print some diagnostics to stderr).\n");
	printf("\t-q|--quiet    \tPrint less (only print events).\n");
	printf("\t-qq           \tPrint nothing (not even events).\n");
	printf("\t--format <fmt>\tPrint using a specified printf-like format\n"
	       "\t              \tstring; read the man page for more details.\n");
	printf("\t--timefmt <fmt>\tstrftime-compatible format string for use with\n"
	       "\t              \t%%T in --format string.\n");
	printf("\t-c|--csv      \tPrint events in CSV format.\n");
	printf("\t-t|--timeout <seconds>\n"
	       "\t              \tWhen listening for a single event, time out "
	       "after\n"
	       "\t              \twaiting for an event for <seconds> seconds.\n"
	       "\t              \tIf <seconds> is 0, inotifywait will never time "
	       "out.\n");
	printf("\t-e|--event <event1> [ -e|--event <event2> ... ]\n"
	       "\t\tListen for specific event(s).  If omitted, all events are \n"
	       "\t\tlistened for.\n");
	printf("\t--execute <program>\n"
	       "\t              \tExecute specified program.  argv[1] will be\n"
	       "\t              \tthe full path of the event.\n\n");
	printf("Exit status:\n");
	printf("\t%d  -  An event you asked to watch for was received.\n",
	       EXIT_SUCCESS );
	printf("\t%d  -  An event you did not ask to watch for was received\n",
	       EXIT_FAILURE);
	printf("\t      (usually delete_self or unmount), or some error "
	       "occurred.\n");
	printf("\t%d  -  The --timeout option was given and no events occurred\n",
	       EXIT_TIMEOUT);
	printf("\t      in the specified interval of time.\n\n");
	printf("Events:\n");
	print_event_descriptions();
}

